src/Build/Logging/ProfilerLogger.cs

```diff
@@ -96,59 +96,165 @@ private void ProjectEvaluationFinishedRaised(object sender, BuildEventArgs e)
                 }
             }
 
-            // Add one single item representing the total aggregated evaluation time for globs
+            if (pruneSmallItems)
+            {
+                // After aggregating all items, prune the ones that are too small to report
+                _aggregatedLocations = PruneSmallItems(_aggregatedLocations);
+            }
+
+            // Add one single top-level item representing the total aggregated evaluation time for globs.
             var aggregatedGlobs = _aggregatedLocations.Keys
                 .Where(key => key.Kind == EvaluationLocationKind.Glob)
-                .Aggregate(new ProfiledLocation(), (profiledLocation, evaluationLocation) => AggregateProfiledLocation(profiledLocation, _aggregatedLocations[evaluationLocation]));
+                .Aggregate(new ProfiledLocation(),
+                    (profiledLocation, evaluationLocation) =>
+                        AggregateProfiledLocation(profiledLocation, _aggregatedLocations[evaluationLocation]));
 
-            _aggregatedLocations[EvaluationLocation.CreateLocationForAggregatedGlob()] = aggregatedGlobs;
+            _aggregatedLocations[EvaluationLocation.CreateLocationForAggregatedGlob()] =
+                aggregatedGlobs;
 
             return new ProfilerResult(_aggregatedLocations);
         }
 
-        private static ProfiledLocation AggregateProfiledLocation(ProfiledLocation location, ProfiledLocation otherLocation)
+        private static void MergeItem(
+            IDictionary<EvaluationLocation, EvaluationLocation> originalLocations,
+            IDictionary<int, int> mergeMap,
+            IDictionary<EvaluationLocation, ProfiledLocation> aggregatedLocations,
+            KeyValuePair<EvaluationLocation, ProfiledLocation> pairToMerge)
+        {
+            ProfiledLocation existingProfiledLocation;
+            if (aggregatedLocations.TryGetValue(pairToMerge.Key, out existingProfiledLocation))
+            {
+                // A previous item, structurally equivalent, is already there
+                // So we aggregate the profiled location times
+                var profiledLocation = AggregateProfiledLocation(existingProfiledLocation, pairToMerge.Value);
+                // We update the *original* key with the aggregated times, so the location table is kept in a sound state regarding parents
+                var originalKey = originalLocations[pairToMerge.Key];
+                aggregatedLocations[originalKey] = profiledLocation;
+                // And we update the merge map to flag that the merge happened
+                mergeMap[pairToMerge.Key.Id] = originalKey.Id;
+            }
+            else
+            {
+                // The item is new, so the profiled times are legit, nothing to aggregate
+                // But we have to check if this item points to a parent that got merged
+                int mergedParent;
+                if (mergeMap.TryGetValue(pairToMerge.Key.Id, out mergedParent))
```

Comment:
Shouldn't this be pairToMerge.Key.ParentId? You are remapping the parent, not the location itself

Reply:
Speaking of, are there tests hitting this?

---
In reply to: [157267787](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157267787) [](ancestors = 157267787)

Reply:
oops :)

---
In reply to: [157268080](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268080) [](ancestors = 157268080,157267787)

Reply:
The merge map potentially forms a chain. Shouldn't you need to follow the chain and use the top parent? I think that only the top parent goes into the AggregatedLocations, right? N locations get merged to one.

---
In reply to: [157268268](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268268) [](ancestors = 157268268,157268080,157267787)

Reply:
They don't form a chain. An item can be merged multiple times, but they all get merged into the first occurrence of the item regarding Ids

---
In reply to: [157268972](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268972) [](ancestors = 157268972,157268268,157268080,157267787)

Reply:
The usual Target evaluation hits this one

---
In reply to: [157268004](https://github.com/Microsoft/msbuild/pull/2813#discussion_r157268004) [](ancestors = 157268004,157267787)